### PR TITLE
Add option to calculate lux value without floating point operations

### PR DESCRIPTION
**Pull request by markus-k**
Many microcontrollers (like ESP32s) do not have a floating point unit, therefore floating point operations have to be done in software. This is slow and consumes quite a lot of flash memory, so it's best avoided if possible.

This PR adds an option to calculate the physical lux value using only integer operations (using u64 for mulitplications internally, though). The resulting value is in milli-lux to cover the full range from 10mlux to 65k lux.

This also moves the calculation of the lux value to a method of `LuxData`, which in turn has to remember the gain and integration-time it was measured with. This allows code generation without floating point operations, as the `lux_phys` would be calculated even if not accessed (this is a guess, but I'd be fairly sure). It is a breaking change, but in my opinion for the better.

Also fixes a sign error in the floating point version of the lux formula.

### CODE REVIEW

File: src/types.rs [hunk 1]
```diff
@@ -1,11 +1,142 @@
...
+            self.gain,
+            self.integration_time,
+        )
+    }
+}
+
+fn raw_to_lux(ch1_data: u16, ch0_data: u16, gain: Gain, itime: IntegrationTime) -> f32 {
+    let ratio = ch1_data as f32 / (ch0_data as f32 + ch1_data as f32);
+    let als_gain: f32 = gain.into();
+    let int_time: f32 = itime.into();
```
Comment (markus-k) [inline src/types.rs]: Might be a good idea to replace these with match expressions as well, but I didn't want to change too much unrelated code in this PR.

File: src/types.rs [hunk 2]
```diff
@@ -1,11 +1,142 @@
...
+        raw_to_lux_u32(
+            self.lux_raw.ch1_raw,
+            self.lux_raw.ch0_raw,
+            self.gain,
+            self.integration_time,
+        )
+    }
+}
+
+fn raw_to_lux(ch1_data: u16, ch0_data: u16, gain: Gain, itime: IntegrationTime) -> f32 {
```
Comment (acalinp) [inline src/types.rs]: Any particular reason to keep this function? As far as I see it, the `raw_to_lux_u32` **feels** to me like it should be faster because of the multiple match expressions (and honestly, how fast do we even need this function to be...)

I suggest to delete `raw_to_lux` and rename the `raw_to_lux_u32` function to simply `raw_to_millilux` and let the user divide the value by 1000 if they truly want it to be in lux.

What do you think?

File: src/types.rs (hunk 1 above)
Comment (acalinp) [reply src/types.rs]: See my comment above. In my opinion, due to your new implementation, the old implementation doesn't bring any advantages. I'd rather have a single good generic implementation than multiple doing the same thing. Plus the user wouldn't even know when to use the `raw_to_lux` and when to use the `raw_to_lux_u32` function.

File: src/types.rs [hunk 3]
```diff
@@ -1,11 +1,142 @@
...
+    } else if (0.45..0.64).contains(&ratio) {
+        ((4.2785 * f32::from(ch0_data)) - (1.9548 * f32::from(ch1_data))) / als_gain / int_time
+    } else if (0.64..0.85).contains(&ratio) {
+        ((0.5926 * f32::from(ch0_data)) + (0.1185 * f32::from(ch1_data))) / als_gain / int_time
+    } else {
+        0.0
+    }
+}
+
+fn raw_to_lux_u32(ch1_data: u16, ch0_data: u16, gain: Gain, itime: IntegrationTime) -> u32 {
```
Comment (acalinp) [inline src/types.rs]: `raw_to_millilux` would be more appropriate, don't you think?

**Review by acalinp — COMMENTED** [review]
Thanks a lot for your contribution! Let's discuss the points I mentioned in the comments. I would really like to simplify to just one function, unless you see any reason we should not?

File: src/types.rs (hunk 2 above)
Comment (markus-k) [reply src/types.rs]: The primary reason I kept this function was backwards compatibility, but since the changes are breaking anyway, we can probably just throw it out. If there are any resolution gains with f32, they are probably way in the noise.

File: src/types.rs (hunk 3 above)
Comment (markus-k) [reply src/types.rs]: Yes, will change that.
